staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go

```diff
@@ -69,6 +69,22 @@ type Encoder interface {
 	Identifier() Identifier
 }
 
+// MemoryAllocator is responsible for allocating memory.
+// It exists to make the cost of object serialization cheaper. In some cases, it allows for allocating memory only once and then reusing it.
+// This approach puts less load on GC and leads to less fragmented memory in general.
+type MemoryAllocator interface {
+	// Allocate reserves memory for n bytes
+	Allocate(n uint64) []byte
+}
+
+// EncoderAllocator  serializes objects in a way that allows callers to manage any additional memory allocations.
+type EncoderAllocator interface {
+	// EncodeWithAllocator writes an object to the provided writer.
+	// In addition, it allows for providing a memory allocator for efficient memory usage during object serialization.
+	// This essentially allows for reusing memory instead of making additional allocations some serializers might require during serialization.
```

Comment:
Maybe switch to sth like:

EncodeWithAllocator writes an object to a stream as Encode does.
However, it may utilize the passed MemoryAllocator to reduce the cost of memory allocations and
garbage collection.